securedrop/tests/client_server_test.py

```diff
@@ -175,7 +176,21 @@ def test_aak_add_contact(self):
                         self.assertEqual("name_v_2", contacts["email_v_2"])
                         self.assertEqual("name_v_3", contacts["email_v_3"])
 
-    def test_aal_login_correct_password_decrypt_contact(self):
+    def test_aal_list_contacts_empty_dictionary(self):
```

Comment:
This is a good test, but we need another where there are other clients (some logged in, some not, some are mutually added contacts, some not) in order to make sure it works. The filter logic you wrote prima facie looks valid, but it still needs testing.

You'll have to create multiple client processes (`multiprocessing.Process` -- threads will not work) that will remain online in the background. `client.main()` takes a filename argument, so you can use a different filename for each so that you can create multiple users, each with their own email/name/contacts. After all client processes are running and logged in in the background, you can log in and list contacts and check to make sure that everything is correct. It is probably easiest to `terminate()` the client processes to clean up, since they'll be in an input loop.

Let me know if you have questions on how to implement this. It may sound confusing.